events/bus.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package events
+
+import (
+	"sync"
+)
+
+// Message contains event info.
+type Message struct {
+	// Name is the name of the event this message was generated from.
+	Name string
+	// Data contains optional event information.
+	Data any
+}
+
+// NewMessage returns a new message with the given name and optional data.
+func NewMessage(name string, data any) Message {
+	return Message{name, data}
+}
+
+// Subscription is a read-only event stream.
+type Subscription struct {
+	id     int
+	value  chan Message
+	events []string
+}
+
+// Message returns the next event value from the subscription.
+func (s *Subscription) Message() <-chan Message {
+	return s.value
+}
+
+// Events returns the names of all subscribed events.
+func (s *Subscription) Events() []string {
+	return s.events
+}
+
+// Bus is used to publish and subscribe to events.
+type Bus struct {
+	subId  int
+	subs   map[int]*Subscription
+	events map[string]map[int]any
+	mutex  sync.RWMutex
```

Comment:
discussion: The scope of this mutex is quite large, and I don't think I agree with this change.  The previous solution was non-blocking and it didn't discard messages (the `default` block in `Bus`.`Publish` here).

I think it is very important that publishing events does not block 'core' database operations, particularly in user-space.  The rapid discarding of events if the buffer fills also far from ideal, and I worry that it will lead to some painful to figure out bug reports, particularly in the P2P space.

Reply:
I don't think you can have non-blocking and also not discarding messages at the same time. You can delay blocking by having a certain size of channel but when it's full we can either block or discard. Unless we have a store to keep track of unsent messages but that is a whole other subject which probably shouldn't apply here. 

The previous solution was blocking as far as I can tell.

Reply:
We have 2 event bus in this PR. We have the internal event bus and the HTTP subscription event bus. The internal event bus will only ever block with the mutex on startup when it adds the internally defined and static subscribers. For the rest of its life, the mutex will never be blocking. Right now the `Publish` method is non-blocking on the channel and that is something we can debate. For internal message exchanges, blocking might actually be desirable.

The HTTP subscription bus, however, has a dynamic list of subscribers. This one will block more often to update the subscriber list. We do not want to block on the channel here that is certain. Since HTTP subscriptions touch user space, using timeouts in the `Publish` method would be appropriate.
```go
for id := range subscribers {
	go func(id int) {
		select {
		case b.subs[id].value <- msg:
			// published event
		case <-time.After(time.Second * 10):
			// channel full, timeout
		}
	}(id)
}
```

Reply:
> The previous solution was blocking as far as I can tell.

It was much less blocking - it was lock free, and the (buffered) `write` channel was consumed by a dedicated Go routine that fed inputs into the (buffered) subscribing channel. And for the write channel to block a subscriber channel would need to be full as well as the write channel.  Atm the write channel is hardcoded to size 0, and the subscribers 100.  It won't throw away data until a subscriber times out after 60 seconds.  We planned on making all of these settings configurable but never got round to it

The proposed solution blocks on every concurrent call regardless of configuration thanks to the quite broad mutex, and there is no way to configure around that (unlike a buffered write channel).  It will discard data as soon as the (size 10) subscriber buffer is filled - this is really not very tolerant at all.

The current solution is far from perfect, but I do think it is better than the proposal.

Reply:
> the mutex will never be blocking

The mutex will block the moment concurrent writers try to write to the bus.

Reply:
Implementation is now channel based like we discussed